vignettes/quitte-data-analysis.Rmd

```diff
@@ -208,24 +126,111 @@ data <- data %>%
 
-`order.levels()` arranges factor levels (e.g. scenario names) in a specific 
-order, thus controlling the order with which items appear in plots.
+
+## Piping
+
+The pipe operator `%>%` is a shortcut to string any number of operations 
+together. So instead of writing a long code worm like
+```{r code worm, eval = FALSE}
+ungroup(mutate(group_by(select(filter(data, 'Reference' == scenario), region, period, value), region), value = cumsum(value)))
+```
+that you would have to read from the inside out, you can write each function 
+call on a separate line, greatly improving readability as data "flows" from top 
+to bottom, each function being applied in succession.
+```{r readable code, eval = FALSE}
+data %>% 
+    filter('Reference' == scenario) %>% 
+    select(region, period, value) %>% 
+    group_by(region) %>% 
+    mutate(value = cumsum(value)) %>% 
+    ungroup()
+```
+
+## Filtering
+`filter()` is a more useful and versatile variant of the data frame extraction 
+operator `[`. Instead of
+```{r extraction, eval = FALSE}
+data[data$period == 2030 & df$scenario != 'Reference',]
+```
+use
+```{r filter, eval = FALSE}
+data %>% 
+    filter(period == 2030,
+           scenario != 'Reference')
+```
+again, improving readability and decreasing proneness to error. Predicates (the
+test) are combined via logical AND, you can use as many of them as you want, 
+and there are lots of helper functions available in the `dplyr` package.
+
+## Selecting
+`select()` selects columns from a data frame. So if you only need columns
+`scenario` and `region`, do
+```{r select A and B, eval = FALSE}
+data %>% 
+    select(scenario, region)
+```
+If you need all but `model`, do
+```{r select all but C, eval = FALSE}
+data %>% 
+    select(-model)
+```
+You can rename columns on the way
+```{r select and rename, eval = FALSE}
+data %>% 
+    select(scenario, year = period)
+```
+
+## Grouping
+`group_by()` subdivides a data frame into groups, such that so-called _window_ 
+functions operate only on the items within a group. If you, for example, want 
+to calculate the cumulated energy production over time for different 
+scenarios and regions, you need to group by those dimensions first.
+```{r grouping, eval = FALSE}
+data %>% 
+    filter(variable == 'PE') %>% 
+    group_by(scenario, region) %>% 
+    mutate(value = cumsum(value)) %>% 
+    ungroup()
+```
+
+## Mutating
+`mutate()` is the low-level function for modifying existing or creating new 
+columns. Its output has the same number of rows as its input.
+```{r mutating, eval = FALSE}
+data %>% 
+    filter(variable == 'PE') %>% 
+    group_by(scenario, region) %>% 
+    mutate(value = cumsum(value)) %>% 
+    ungroup()
+```
+
+## Summarising
+`summarise()` is the low-level function for aggregating variables. Its output 
+has _one_ row for every group. All ungrouped columns are dropped.
+```{r summarising, eval = FALSE}
+data %>% 
+    group_by(scenario) %>% 
+    summarise(value = mean(value)) %>% 
```

Comment:
```suggestion
    summarise(value = mean(value), .groups = 'drop') %>% 
```

Reply:
Same as calling `ungroup()`.  Both are strictly not necessary since the default behaviour is "drop_last" and we have only one group, but this also hints of the existence of `ungroup()` for those who did not read the tidyverse tutorial.